File: 2020/day18/sol.cpp
#include <fstream>
#include <iostream>
#include <queue>
#include <stack>
#include <vector>

using u64 = uint64_t;

std::vector<std::string> tokenize(std::string expr) {
    std::vector<std::string> out = {};

    u64 pos = 0;
    u64 last = 0;
    while (pos < expr.length()) {
        if (expr[pos] == '(') {
            int paren = 1;
            while (paren != 0) {
                pos++;
                if (expr[pos] == '(') {
                    paren++;
                } else if (expr[pos] == ')') {
                    paren--;
                }
            }
        }
        if (expr[pos] == ' ') {
            out.push_back(expr.substr(last, pos - last));
            last = pos + 1;
        }
        pos++;
        if (pos >= expr.length()) {
            out.push_back(expr.substr(last));
        }
    }

    return out;
}

std::string remove_parens(const std::string& s) {
    return s.substr(1, s.length() - 2);
}

u64 eval(std::string line) {
    auto tokens = tokenize(line);
    auto& first_token = tokens[0];
    u64 acc = [first_token]() -> u64 {
        if (first_token[0] == '(') {
            return eval(remove_parens(first_token));
        } else {
            return std::stoul(first_token);
        }
    }();

    for (size_t i = 1; i < tokens.size(); i += 2) {
        auto op = tokens[i];
        auto value = [i, &tokens]() -> u64 {
            auto& expr = tokens[i + 1];
            if (expr[0] == '(') {
                return eval(remove_parens(expr));
            } else {
                return std::stoul(expr);
            }
        }();

        if (op == "+") {
            acc += value;
        } else {
            acc *= value;
        }
    }

    return acc;
}

std::vector<std::string> split_string(std::string s, const std::string delim = " ", std::vector<std::string> acc = {}) {
    auto pos = s.find(delim);
    if (pos != std::string::npos) {
        acc.push_back(s.substr(0, pos));
        return split_string(s.substr(pos + delim.length()), delim, acc);
    } else {
        acc.push_back(s);
        return acc;
    }
}

u64 eval_with_precedence(std::string line) {
    // shunting yard algorithm
    std::stack<std::string> operators;
    std::queue<std::string> output;

    auto tokens = split_string(line);
    for (auto token : tokens) {
        if (token.find_first_not_of("0123456789") == std::string::npos) { // if the token is a number
            output.push(token);
        } else if (token[0] == '+' || token[0] == '*') {
            while (!operators.empty() && operators.top() == "+") {
                output.push("+");
                operators.pop();
            }
            operators.push(token);
        } else if (token[0] == '(') {
            auto digit_position = token.find_first_of("0123456789");
            for (size_t i = 0; i < digit_position; ++i) {
                operators.push("(");
            }
            output.push(token.substr(digit_position));
        } else if (token.find(')') != std::string::npos) {
            auto first_bracket_pos = token.find_first_not_of("0123456789");
            output.push(token.substr(0, first_bracket_pos));
            for (size_t i = first_bracket_pos; i < token.length(); ++i) {
                while (operators.top() != "(") {
                    output.push(operators.top());
                    operators.pop();
                }
                operators.pop();
            }
        }
    }

    while (!operators.empty()) {
        output.push(operators.top());
        operators.pop();
    }

    std::stack<u64> execution;
    while (!output.empty()) {
        if (output.front() == "+" || output.front() == "*") {
            auto val1 = execution.top();
            execution.pop();
            auto val2 = execution.top();
            execution.pop();

            if (output.front() == "+") {
                val2 += val1;
            } else {
                val2 *= val1;
            }

            execution.push(val2);
        } else {
            execution.push(std::stoull(output.front()));
        }
        output.pop();
    }

    return execution.top();
}

int main() {
    std::ifstream input("input");
    if (!input.is_open()) {
        std::cerr << "Can't read puzzle input file.\n";
        return -1;
    }

    u64 sum1 = 0;
    u64 sum2 = 0;
    std::string line;
    while (std::getline(input, line)) {
        sum1 += eval(line);
        sum2 += eval_with_precedence(line);
    }

    std::cout << "part 1: " << sum1 << '\n';
    std::cout << "part 2: " << sum2 << '\n';

    return 0;
}

File: 2020/day18/sol.rs
use std::collections::VecDeque;
use std::fs;

fn tokenize(expr: &str) -> Vec<String> {
    let mut ret = vec![];

    let mut pos = 0;
    let mut last = 0;

    while pos < expr.len() {
        if expr.chars().nth(pos) == Some('(') {
            let mut paren = 1;
            while paren != 0 {
                pos += 1;
                if expr.chars().nth(pos) == Some('(') {
                    paren += 1;
                } else if expr.chars().nth(pos) == Some(')') {
                    paren -= 1;
                }
            }
        }
        if expr.chars().nth(pos) == Some(' ') {
            ret.push(expr.chars().skip(last).take(pos - last).collect());
            last = pos + 1;
        }
        pos += 1;
        if pos >= expr.len() {
            ret.push(expr.chars().skip(last).collect());
        }
    }

    ret
}

fn remove_parens(s: &str) -> String {
    s.chars().skip(1).take(s.len() - 2).collect()
}

fn eval(line: &str) -> usize {
    let tokens = tokenize(line);
    let first_token = &tokens[0];

    let mut acc = if first_token.starts_with('(') {
        eval(&remove_parens(first_token))
    } else {
        first_token.parse::<usize>().unwrap()
    };

    for i in (1..tokens.len()).step_by(2) {
        let op = &tokens[i];
        let value = {
            let expr = &tokens[i + 1];
            if expr.starts_with('(') {
                eval(&remove_parens(expr))
            } else {
                expr.parse::<usize>().unwrap()
            }
        };

        if op == "+" {
            acc += value;
        } else {
            acc *= value;
        }
    }

    acc
}

fn eval_with_precedence(line: &str) -> usize {
    let mut operators: Vec<String> = vec![];
    let mut output: VecDeque<String> = VecDeque::new();

    let tokens = line.split_whitespace().collect::<Vec<_>>();
    for token in tokens {
        if token.chars().all(|c| c.is_ascii_digit()) {
            output.push_back(token.to_owned());
        } else if token.starts_with('+') || token.starts_with('*') {
            while !operators.is_empty() && operators.last() == Some(&"+".to_string()) {
                output.push_back("+".to_owned());
                operators.pop();
            }
            operators.push(token.to_owned());
        } else if token.starts_with('(') {
            let (digit_position, _) = token
                .chars()
                .enumerate()
                .find(|(_, c)| c.is_ascii_digit())
                .unwrap();

            for _ in 0..digit_position {
                operators.push("(".to_owned());
            }

            output.push_back(token.chars().skip(digit_position).collect());
        } else if token.contains(')') {
            let (first_bracket_pos, _) = token
                .chars()
                .enumerate()
                .find(|(_, c)| !c.is_ascii_digit())
                .unwrap();
            output.push_back(token.chars().take(first_bracket_pos).collect());
            for _ in first_bracket_pos..token.len() {
                while operators.last() != Some(&"(".to_string()) {
                    output.push_back(operators.pop().unwrap().to_string());
                }
                operators.pop();
            }
        }
    }

    while !operators.is_empty() {
        output.push_back(operators.pop().unwrap().to_string());
    }

    let mut execution = vec![];
    while !output.is_empty() {
        if let Some(operation) = output.front() {
            if operation == "+" || operation == "*" {
                let val1: usize = *execution.last().unwrap();
                execution.pop();
                let mut val2 = *execution.last().unwrap();
                execution.pop();

                if operation == "+" {
                    val2 += val1;
                } else {
                    val2 *= val1;
                }

                execution.push(val2);
            } else {
                execution.push(output.front().unwrap().parse::<usize>().unwrap());
            }
            output.pop_front();
        }
    }

    execution.pop().unwrap()
}

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let part1 = input.lines().map(eval).sum::<usize>();
    let part2 = input.lines().map(eval_with_precedence).sum::<usize>();

    println!("part 1: {}", part1);
    println!("part 2: {}", part2);
}
